qt/src/hocr/HOCRDocument.cc

```diff
@@ -82,6 +82,54 @@ bool HOCRDocument::editItemAttribute(const QModelIndex& index, const QString& na
 	return true;
 }
 
+// Might be more logical to accept oldParent and oldRow as parameters rather than itemIndex
+// (since we need them anyway), if all our callers are likely to know them. Swap does; drag&drop might not.
+QModelIndex HOCRDocument::moveItem(const QModelIndex& itemIndex, const QModelIndex& newParent, int newRow) {
+	HOCRItem* item = mutableItemAtIndex(itemIndex);
+	if(!item) {
+		return QModelIndex();
+	}
+	QModelIndex ancestor = newParent;
+	while(ancestor.isValid()) {
+		if(ancestor == itemIndex) {
+			return QModelIndex();
+		}
+		ancestor = ancestor.parent();
+	}
+	int oldRow = itemIndex.row();
+	QModelIndex oldParent = itemIndex.parent();
+	bool decr = false;
+	if(decr = (oldParent == newParent && oldRow < newRow)) {
+		--newRow;
+	}
+	HOCRItem* parentItem = mutableItemAtIndex(newParent);
+	if(parentItem) {
+		beginRemoveRows(oldParent, oldRow, oldRow);
+		takeItem(item);
+		endRemoveRows();
+		beginInsertRows(newParent, newRow, newRow);
+		parentItem->insertChild(item, newRow);
+		endInsertRows();
+	} else if(item->itemClass() == "ocr_page") {
+		beginRemoveRows(QModelIndex(), oldRow, oldRow);
+		HOCRPage* pageItem = m_pages.takeAt(oldRow);
+		endRemoveRows();
+		beginInsertRows(QModelIndex(), newRow, newRow);
+		m_pages.insert(newRow, pageItem);
```

Comment:
This (and `m_pages.takeAt` above) won't re-assign the page indices. You should be able to just use `takeItem` instead of `m_pages.takeAt`, and if you add a generic `HOCRDocument::insertItem` you can actually use the same code regardless of whether the item is a page or not. The only page-item specific part remains the `emit dataChanged`.

Reply:
Ah, so it doesn't. Fixed, and insertItem (private, analogous to take) added. I believe that as insert and take both emit their own dataChanged for pages, it actually isn't necessary for moveItem to do so—the only page logic left is checking whether an invalid parent is OK.

Am I correct in assuming that hOCR id attributes are subordinate to the tree structure in determining exported content flow, and do not need to be updated?

Reply:
Id's are just there to be unique but don't play any other role. No need to change them.